src/shared/bootspec.c

```diff
@@ -1046,6 +1052,189 @@ int boot_config_augment_from_loader(
+        if (strcmp(preprelast, "block"))
+                return -ENOENT;
+
+        aprelast = prelast;
+
+        snprintf(sysfsname, sizeof(sysfsname) - 1, "/sys/dev/block/%d:%d", major(b), minor(b));
+        r = readlink(sysfsname, bufb, sizeof(bufb) -1);
+        if (r < 0)
+                return r;
+
+        /* bufb is in the form of '../../devices/pci0000:00/0000:00:01.2/0000:01:00.1/ata4/host3/target3:0:0/3:0:0:0/block/sde/sde3' */
+        split_last_three_by_slash(bufb, &preprelast, &prelast, &last);
+        if (strcmp(preprelast, "block"))
+                return -ENOENT;
+
+        return strcmp(prelast, aprelast);
+}
+
+static int check_part_type_duplicates_blkid(const char *part_uuid, const char *dos_part_uuid, dev_t *devtptr) {
+#if HAVE_BLKID
+        _cleanup_(blkid_dev_iterate_endp) blkid_dev_iterate iter = NULL;
+        blkid_cache cache = NULL;
+        blkid_dev dev;
+        int count = 0;
+        int r;
+
+        r = blkid_get_cache(&cache, NULL);
+        if (r < 0)
+                return log_error_errno(r, "Failed to initialize blkid cache");
+
+        blkid_probe_all(cache);
+        iter = blkid_dev_iterate_begin(cache);
+        while (blkid_dev_next(iter, &dev) == 0) {
+                _cleanup_(blkid_free_probep) blkid_probe b = NULL;
+                const char *v;
+                const char *devname;
+
+                dev = blkid_verify(cache, dev);
+                if (!dev)
+                        continue;
+
+                devname = blkid_dev_devname(dev);
+
+                errno = 0;
+                b = blkid_new_probe_from_filename(devname);
+                if (!b)
+                        return log_error_errno(errno ?: SYNTHETIC_ERRNO(ENOMEM), "Failed to open file system \"%s\": %m", devname);
+
+                blkid_probe_enable_superblocks(b, 1);
+                blkid_probe_set_superblocks_flags(b, BLKID_SUBLKS_TYPE);
+                blkid_probe_enable_partitions(b, 1);
+                blkid_probe_set_partitions_flags(b, BLKID_PARTS_ENTRY_DETAILS);
+
+                errno = 0;
+                r = blkid_do_safeprobe(b);
+
+                /*
+                 * Ignore the error because the goal is only to find (if any) partition with
+                 * the same GUID
+                 */
+                r = blkid_probe_lookup_value(b, "PART_ENTRY_TYPE", &v, NULL);
+                if (r != 0)
+                        continue;
+
+                if (devtptr) {
+                        dev_t devid = blkid_probe_get_devno(b);
+                        if (check_parent_by_devid(devid, *devtptr))
+                                continue;
+                }
+
+                if (streq(v, part_uuid))
+                        count++;
+
+                if (dos_part_uuid && streq(v, dos_part_uuid))
+                        count++;
+        }
+
+        return count;
+#else
+        /* we don't have blkid, return a positive result */
+        return 1;
+#endif
+}
```

Comment:
quite frankly, drop the blkid based code path entirely. This is in a high-level, user-facing tool, i.e. udev, and is about generating a friendly warning. It's entirely fine if this only works once udev is up. 

We have blkid-based codepaths as alternative in some cases, but that's only for us in certainy very specific cases during early boot where we really need the functionality but udev is known not to be up yet. But this is noe one of those cases. Hence, please drop this check entirely, please

Reply:
ok